Add New-EditorFile

tylerl0706 opened the pull request:
Here's what you can do:

```powershell
Open-EditorFile foo.txt -Force
```

This will create a new file called `foo.txt` if no file exists. It will throw an error if you don't add `-force` and `foo.txt` does not exist.

Also,
```powershell
Get-Process | Open-EditorFile foo.txt -Force
```

This will create a new file called `foo.txt` with the contents of what came in through the pipeline. Another example:

```powershell
irm https://raw.githubusercontent.com/PowerShell/PowerShell/master/tools/install-powershell.ps1 | psedit install-powershell.ps1 -Force
```

You can also do:
```powershell
Get-Process | Open-EditorFile
```
This will open an Untitled file in your editor and put the contents in it.

This works in local and remote sessions.

Note, Get-Process has an issue in remote sessions when piping to a file. I originally thought it was my code but it was not:
https://github.com/PowerShell/PowerShell/issues/6115

The vscode-powershell PR is here: https://github.com/PowerShell/vscode-powershell/pull/1197

Resolves https://github.com/PowerShell/PowerShellEditorServices/issues/414
Resolves https://github.com/PowerShell/PowerShellEditorServices/issues/413

daviwil commented on module/PowerShellEditorServices/Commands/Public/CmdletInterface.ps1:
```diff
@@ -78,10 +78,38 @@ function Unregister-EditorCommand {
+        $FilePaths,
+
+        [Parameter(ValueFromPipeline=$true)]
+        $NewValue,
```
[Set-Content](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.management/set-content?view=powershell-5.1) uses `Value` as the name for this parameter, maybe good to be consistent since the behavior is similar?  Not really a big deal, though.

daviwil commented on module/PowerShellEditorServices/Commands/Public/CmdletInterface.ps1:
```diff
@@ -78,10 +78,38 @@ function Unregister-EditorCommand {
+    end {
+        $FilePaths | ForEach-Object {
+            if (-not (Test-Path $_) -and $Force) {
+                $container > $_
```
Definitely would be interested in Keith's feedback about this vs `Set-Content -Force`

rkeithhill replied:
Yeah, in the context of file manipulation the Force parameter is usually associated with dealing with read only files.

I'm also a bit confused about the purpose of this command.  When I see the name `Open-EditorFile` I don't associate that with setting an initial content (NewValue).  Isn't there a `New-EditorFile` which would take an optional Value (as well as Force to overwrite existing files)?  

I think this command is trying to do a bit too much.  Seems like it should just open the specified file in the editor workspace.  Folks can then use the $pseditor api to modify its contents.

tylerl0706 replied:
@rkeithhill `New-EditorFile` does not exist which is why I bundled it with `Open-EditorFile` but I guess we can just have another function.

rkeithhill replied:
I think that would be cleaner and more PowerShelly. :-)

rkeithhill commented on module/PowerShellEditorServices/Commands/Public/CmdletInterface.ps1:
```diff
@@ -78,10 +78,38 @@ function Unregister-EditorCommand {
+    param(
+        [Parameter(Mandatory=$true)]
+        [ValidateNotNullOrEmpty()]
+        $FilePaths,
```
This should be singular and in this case I think `Path` is sufficient and consistent with the other PowerShell file manipulation commands.  

Also, I think it would be better to have this command accept pipeline input for the path instead of content.  But that requires some changes to the parameter config - see the Examples\PathProcessingWildcards.ps1 for an example of how to do that. 

With this change, I would be able to do:
```posh
gci *.psd1 -r | Open-EditorFile
```
That seems more useful to me than using the pipeline to overwrite (or create) files with new content.

SeeminglyScience replied:
Could do both:

```powershell
function Open-EditorFile {
    [CmdletBinding(PositionalBinding=$false, DefaultParameterSetName='Path')]
    param(
        [Parameter(Mandatory, ParameterSetName='Path', ValueFromPipeline, ValueFromPipelineByPropertyName, Position=0)]
        [Parameter(Mandatory, ParameterSetName='Value', Position=0)]
        [ValidateNotNullOrEmpty()]
        [Alias('FullName')]
        [string] $Path,

        [Parameter(ParameterSetName='Value', ValueFromPipeline)]
        [Parameter(ParameterSetName='Path')]
        [ValidateNotNullOrEmpty()]
        [string] $Value
    )
    begin {
		if ($MyInvocation.ExpectingInput) {
            $valueList = [System.Collections.Generic.List[string]]::new()
        }
    }
    process {
        if ($PSCmdlet.ParameterSetName -eq 'Value') {
            $valueList.Add($Value)
            return
        }

        $psEditor.Workspace.OpenFile($Path)
        if ($Value) {
            $psEditor.GetEditorContext().CurrentFile.InsertText($Value)
        }
    }
    end {
        if ($PSCmdlet.ParameterSetName -eq 'Path') {
            return
        }

        $psEditor.Workspace.OpenFile($Path)
        $psEditor.GetEditorContext().CurrentFile.InsertText($valueList -join [Environment]::NewLine)
    }
}
```

Allows for `gci *.ps1 | Open-EditorFile -Value $x` and `gc .\sourceFile.ps1 | Open-EditorFile destinationFile.ps1`.

tylerl0706 commented:
I tried SO hard to get C# to read in a psm1 as a string... but eventually gave up and put the whole module in a string instead...

The benefit of this approach is that we can add more functions to the module string and they will be automatically added to remote sessions.

SeeminglyScience commented on src/PowerShellEditorServices/Session/RemoteFileManager.cs:
```diff
@@ -36,55 +36,161 @@ public class RemoteFileManager
+            $psedit = New-Module -ScriptBlock ([Scriptblock]::Create($PSEditModule)) -Name PSEdit
+            $psedit.ExportedFunctions.Keys | ForEach-Object {
+                Set-Item -Path function:\global:$_ -Value $asdf.ExportedFunctions[$_].Definition -Force
+            }
```
If we're going the module route, can we pipe this into `Import-Module` instead?

tylerl0706 replied:
@SeeminglyScience I tried `Import-Module` initially but it did not work. None of the cmdlets would show up... Could I have been missing a particular flag? I tried `-Force` and `-Scope`.

SeeminglyScience replied:
@tylerl0706 try `-Global`.  I thought you only needed that when importing from a different session state, but I guess it's scope based. TIL

SeeminglyScience commented on src/PowerShellEditorServices/Session/RemoteFileManager.cs:
```diff
@@ -36,55 +36,161 @@ public class RemoteFileManager
-            {
-                Remove-Item -Path 'function:\global:Open-EditorFile' -Force
+            Get-Command | Where-Object {$_.Source -eq 'PSEdit'} | ForEach-Object {
+                Remove-Item -Path function:\global:$($_.Name) -Force
```
Same as above,  `Get-Module PSEdit | Remove-Module` would be preferable

tylerl0706 commented:
Ok I've addressed all of @SeeminglyScience's feedback.

I've also added a small additional feature to this... You can now do:
```powershell
Get-Process | Open-EditorFile
```

This will open an Untitled file in your editor and put the contents in it.

This works in local and remote sessions.

tylerl0706 commented:
@daviwil this PR is substansially different than when you signed off. Can you look again?

Also, @rkeithhill, I've addressed your feedback.

rkeithhill commented on module/PowerShellEditorServices/Commands/Public/CmdletInterface.ps1:
```diff
@@ -77,13 +77,98 @@ function Unregister-EditorCommand {
+    param(
+        [Parameter(Mandatory=$true, ValueFromPipeline=$true)]
+        [ValidateNotNullOrEmpty()]
+        $Path
```
As long as this parameter supports this scenario `Get-ChildItem *.ps1 -r -file | Open-EditorFile`, I'm good.

tylerl0706 replied:
So @rkeithhill ... as it turns out, this will proceed to open the file temporarily (with the italicized file name)... so what ends up happening is, it opens the file temporarily and then the next file opens and replaces the file temporarily.

I'm not sure how to durably open files - if that's possible. I'll have to dig around.

@daviwil any thoughts?

rkeithhill replied:
Ah, yes.  That's why I double-click the editor window tab.  That converts it from a temp (resuable) window to a permanent one.  Not sure how you do that programmatically though.

tylerl0706 replied:
Exactly. Maybe if @SeeminglyScience or @daviwil know? Otherwise, can I convince you to call this out of scope 😄

rkeithhill replied:
You can use the `vscode.window.showTextDocument()` that takes a TextDocumentShowOptions object. One of the fields of that object is `preview?`.  That should allow you to create each window in a "non-preview" window.  

https://code.visualstudio.com/docs/extensionAPI/vscode-api#_a-nametextdocumentshowoptionsaspan-classcodeitem-id239textdocumentshowoptionsspan

SeeminglyScience replied:
Eh... kind of raises more questions though.

1. Should there be a `isPreview` overload to `OpenFile`?

1. Does that translate well to other editors?

1. Should it just be the default?

1. Should we instead create a request for the `workbench.action.keepEditor` VSCode command and have this function send that request only if the host is VSCode?

tylerl0706 replied:
I had this thought too. I have a working version that sends `preview` along with the `OpenFile` message but I'm not sure that's the best approach as new editors pop up with different functionalities.

rkeithhill replied:
I would expect that if I used this command to open three files (however I specify those three files) that there would be three editor windows open when the command completes (not one editor window with the last file in it).  For now, I think we just specify isPreview as false.  If asked, we could add preview as a parameter later.  I kind of doubt it will be needed.

tylerl0706 replied:
@rkeithhill so you're saying that the change then is in vscode-powershell to always set the isPreview to false?

That's a quick fix and doesn't require a whole lot of extra plumbing. I'm ok with that, if I'm understanding you correctly.

rkeithhill replied:
That would be my preference.  Later we could update the openFile() method to take an optional "preview" Boolean parameter.

tylerl0706 replied:
> Later we could update the openFile() method to take an optional "preview" Boolean parameter.

What if I already have this working? Is it worth adding it now or saving it for when someone actually asks for that?

tylerl0706 replied:
I'll just push and you can judge whether we should keep it in or not

rkeithhill replied:
I'd like to have that option, so yeah, if it's working - go for it.

tylerl0706 replied:
@rkeithhill pushed that. Here's the vscode-powershell PR as well:
https://github.com/PowerShell/vscode-powershell/pull/1197

tylerl0706 commented:
Ah! This resolves this as well:
https://github.com/PowerShell/PowerShellEditorServices/issues/413

tylerl0706 commented:
oh... appveyor.

tylerl0706 commented:
Can you guys give a thumbs up to this if you still approve? There have been some substantial (enough) changes that those approvals don't really stand anymore.
